include/util/exception.hpp

```diff
@@ -54,6 +56,219 @@ class exception final : public std::exception
     virtual void anchor() const;
     const std::string message;
 };
+
+/**
+ * Indicates a class of error that occurred that was caused by some kind of
+ * external input (common examples are out of disk space, file permission errors,
+ * user supplied bad data, etc).
+ */
+class runtime_error : public std::runtime_error
+{
+    using Base = std::runtime_error;
+
+  protected:
+    explicit runtime_error(const char *message,
+                           const std::string &sourceref,
+                           const std::string &code_description)
+        : Base(message), sourceref(sourceref)
+    {
+        messagebuf = code_description + ": " + Base::what() + " (at " + sourceref + ")";
```

Comment:
Why `Base::what()` here? Won't this be always empty since we're in the progress of creating an exception here?